---
title: "R Notebook"
output: html_notebook
---

__Name:__ Mitali Singh

__Student ID:__ 999290264

For each of the exercises below, take advantage of R markdown and the notebook features.  That is include the code and have Rstudio embed your answers.  See Exercise 0 for an example


## Exercise 0 (example)
_Provide the name of your current working directory and list the files therein_
```{r}
getwd()
dir()
```

(You would then click on the > arrow in your answer code block to run it before continuing)

## Exercise 1

_Use sum() to determine the sum of numbers from 2000 to 20000.  Provide your code and the actual sum in your answer._

```{r}
sum(2000,20000)
```

## Exercise 2
_In one or two sentences, describe what the above code snippet it did._

> 5 is assigned to a variable 'a' and a sequence of numbers from 2 to 20 are assigned to another variable 'b'. The command is 3rd and 4th line simply reads the content of variable a and b respectively.

## Exercise 3
_Add the contents of a and b together and place the results in a new object.  Examine the result.  Include your code.  Try using both `sum()` and `+`; do you get different results?  If so, why?_

```{r}
a <- 5 
b <- 2:20
sum <- sum(a,b)
sum
```
> This command sums up all the numbers in both variables together (2-20 and 5) and gives us one single result.
 
```{r}
sum2 <- a+b
sum2
```
> Whereas the command with '+' runs a loop of adding the element of variable a (5) to the elements of variable b (2-20) one at a time. So, it adds 5 to 2 first and then 5 to 3 and so on. 

## Exercise 4
_What is the sum of the 5th through 10th element of object b?  Provide your code and the sum._
```{r}
sum (b [c(5:10)])
```


## Exercise 5
_What is the sum of the 3rd, 8th, and 10th element of b?  For both of these exercises should only need to have "b" in your code once._

```{r}
sum (b[c(3,8,10)])
```


## Exercise 6
_When extracting from a 2-dimensional object, which number specifies rows and which specifies columns?_

_What does `m[3,]` do?_

_How can you extract the 3rd, 4th and 5th columns of m together as one object?_


```{r}
m <- matrix(data=1:25, ncol=5, byrow=T)
m
m[5,5]
```
>  The first number identifies row and the second number identifies column

```{r}
m[3,]
```

>  ```m[3,]``` displays all the elements of row 3 (includes all columns for row 3)
```{r}
m[,3:5]
```



## Exercise 7
_What does the cbind command do?  How about rbind?_
```{r}
cbind(m,101:105)

```
> cbind takes a sequence of vectors,matrix or data frame arguments and combines by columns


```{r}
rbind(m,101:105)
```
> rbind takes a sequence of vectors,matrix or data frame arguments and combines by rows

_Create a new object "n" where the first row is a new row of numbers (your choice) and the following rows are the followed by the m matrix._
```{r}
n <- rbind(2:6,m)
n
```


_Want more? (OPTIONAL) do the same but reverse the order of the rows from m._


## Exercise 8


```{r}
fly.worm <- read.delim("~/Downloads/fly_worm_blast.out.gz",header=F)
head(fly.worm)
summary(fly.worm)
```


```{r}
colnames(fly.worm) <-c("query_id", "subject_id", "pct_ident", "len", "mis", "gaps", "gb", "qe", "sb", "se", "E", "Score")
head(fly.worm)
summary(fly.worm)
Evalue <- fly.worm[,11]
Evalue==0
table(Evalue==0)

Percentid <- fly.worm[,3]
Percentid>50
table(Percentid>50)
table (Evalue==0&Percentid<50)

```

_How many hits have an e-value of 0?_
> 2443 hits have an e-value of 0

```{r}
table(Evalue==0)
```


_How many have hits have a percent identity > 50?_
```{r}
table(Percentid>50)
```

> 5271 hits have a percent identity >5

_Recalculate the above values but in percentage of hits rather than absolute values._

> for E_value=0: 

```{r}
2443/(32431+2443)*100
```

> For percent identity>50:

```{r}
5271/(29603+5271)*100
```

 _How many hits have an E-value of 0 and have a percent identity less than 50?_

```{r}
table (Evalue==0&Percentid<50)
```

> 948 hits have an E-value of 0 and a percent identity <50

_What is the minimum percent identity of the hits with an E-value equal to 0?_
```{r}

min(fly.worm[which(fly.worm["E"]==0),"pct_ident"])
```
> The minimum percent identity of the hits with an E-value to 0 is 24.07


## Excercise 9
_Are you surprised that sequences with relatively low percent identity can still have an E-value of 0?_
>  yes

_State a hypothesis about what alignment properties might produce a zero E-value even when the percent identity is less than 50%._

> The low E-value in low percent identity alignments could be due to the fact that the few amino acids that actually align do so with same amino acids(alanine with alanine, valine with valine, etc.) i.e. no/very few mismatches.

_Test your hypothesis_
```{r}
allEval<- fly.worm[which(fly.worm["E"]==0),]
allEval[which(allEval["pct_ident"]<50),]
allEval[which(allEval["pct_ident"]>50),]
```
> The alignments with low E-score and low percent identity (<50) showed to have higher number of mismatches and gaps than the alignments with low E-score and high percent identity. This result proves my hypothesis wrong. Therefore it can be concluded that low E-value is not indicative of less number of mismatches. .


## Excercise 10
_It sometimes is useful to create a random subset of a large data set like the fly.worm results.  Create a new dataframe `fly.worm.small` that has 1,000 rows taken at random from the larger data set._

```{r}
fly.worm[sample(nrow(fly.worm),1000),]
fly.worm.small <-fly.worm[sample(nrow(fly.worm),1000),]
```

## Exercise 11
_Use the `summary()` and `head()` functions to examine the stereotypes object._
```{r}
stereotype <-read.csv("~/Downloads/stereotypes.csv",header=T)
summary(stereotype)
head(stereotype)
```

_Record the answers to the "activity break sections" from Ciera's visualiztio tutorial here:_ 


 **_Activity 1_**
 
>  1) Nerd females binge drink coffee but hardly ever drink alcohol
 
```{r}
nerds_and_metalheads <- subset(stereotype, population == "nerd" | population == "metalhead")

males.only<- subset(stereotype, gender == "male")
males.metalheads <- subset(males.only, population == "metalhead"& beer > 25& tacos > 20)
males.hipsters <- subset(males.only, population == "hipster"& beer > 25& tacos > 20)
nrow(males.metalheads)
nrow(males.hipsters)
```

> 2) Yes, there are 64  more metalheads who binge drink alcohol and eat >20 tacos in a week.

> 3) Are there more males in the metalheads who binge coffee (over 12 a week) and eat over 8 tacos a week in the metal heads or females metalheads?

```{r}
females.only<- subset(stereotype, gender == "female")
females.metalheads <- subset(females.only, population == "metalhead"& coffee > 12& tacos > 8)
males.metalheads <- subset(males.only, population == "metalhead"& coffee > 12& tacos > 8)
nrow(females.metalheads)
nrow(males.metalheads)
```
Yes, there are 35 more male metalheads than female metalheads who binge drink coffee (over 12 a week) and eat over 8 tacos a week.

> 4) the '==' function means that the query is **exactly** equal to the input. Example: E-value==0 will look for E-values that are exactly equal to 0, nothing else around it

> Activity 2 : VISUALIZATION

1) View the distribution of showers. Color by population.

```{r}
pl4<- ggplot(data=stereotype, mapping=aes(x=shower,fill=population))
pl4<-pl4+geom_histogram()
pl4
```

2) View the distribution of beer consumption and color by gender.

```{r}
pl5<- ggplot(data=stereotype, mapping=aes(x=beer,fill=gender))
pl5<-pl5+geom_histogram()
pl5
```

3)Find a way to create a plot using the the geom violin with the stereotype data set


> ACTIVITY 3

1) Make a graph that visually describes the difference between beer consumption between gender in the nerd category

```{r}
nerds<- subset(stereotype, population=="nerd")
qplot(beer, data = nerds, geom = "density", fill = gender)

```



2) Using the ggplot2 documentation, explore three different geoms, using either the full data set or a subset of the data. Tell me something interesting about the data.

```{r}
library(ggplot2)
qplot(computer, data=stereotype, geom="density", fill=population, alpha=I(0.5))
```

```{r}
hipstersonly<-subset(stereotype, population="hipster")
qplot(computer,data=hipstersonly, geom="bar", fill=gender)
```

```{r}

p0<-ggplot(data=hipstersonly, mapping=aes(x=computer, y=age,fill=gender))
p0<-p0+geom_point()+geom_smooth()
p0
```
> This data indicates that within the entire population, hipster females between 20-23 years of age spend the most time at their computers. 

3) 

> Activity 4

1) With the p base layer, make a unique plot by adding additional layers.

```{r}
p <- ggplot(stereotype, mapping=aes(x=population,y=shower, fill = population))
p<-p+geom_boxplot()
p+coord_flip()
p
```

2) Make new base layer. Explore what the difference is between color = and fill =. Use your own words to discern the difference
```{r}
q<-ggplot(data=stereotype, mapping=aes(x=age, fill=gender, binwidth=4))
q<-q+geom_histogram()
q
```
```{r}
m<-ggplot(data=stereotype, mapping=aes(x=age, color=gender, binwidth=4))
m<-m+geom_histogram()
m
```
> The ```fill=``` command fills in between lines/boundaries of the individual bargraphs with a different color for each component. On the other hand, the ```color``` command puts different colors on the outlines of the graphs(only the border), it doesn't fill in within the boundaries.

> Activity 5

1) Read about and use the following modifiers correctly with the any portion of the sterotypes dataset you want.
```{r}
ggplot(stereotype, aes(beer, tacos, color = population, shape = gender)) + 
geom_point()
```

2) Modify color with scale_color_hue()

```{r}
n<-ggplot(stereotype, aes(beer, tacos, color = population, shape = gender))+ geom_point()
n<- n+ scale_color_hue( c = 40, l = 50)
n
```

3)Make a tricked out box plot using the stereotypes data. Use at least 3 of the aesthetic options available for geom_boxplot.

```{r}
b<- ggplot(stereotype,aes(beer,coffee, color=population))+geom_boxplot(outlier.color="black", outlier.shape=1, position="dodge", varwidth=TRUE)
b
```


## Exercise 12
_Use ggplot to explore the relationship between score ("Score") and alignment length("len").  Provide a plot that illustrates the relationship and describe the relationship in words._

```{r}
library (ggplot2)
data=fly.worm
pl1<-ggplot(data=fly.worm, mapping=aes(x=len, y=Score))
pl1 <- pl1 + geom_smooth() 
pl1
```



## Excercise 13
_While you might expect that BLAST results with long alignments would have high scores, this is not always the case.  Form a hypothesis as to what might influence the relationship between alignment length and score.  Use ggplot to make a new plot to explore this hypothesis.  Does the plot support your hypothesis or not?  State your hypothesis, provide the code for your plot, and state your conclusion._

> BLAST results with long alignments do not always have high scores because the alignments could have lots of mismatches. The plot supports my hypothesis as its shows that the length of an alignment and number of mismatches have a linear relationship which proves that an alignment of long length usually has more mismatches which would give a lower score. 
Conclusion: Long alignments do not always have high scores due to the presence of high number of mismatches. 

```{r}
pl3<-ggplot(data=fly.worm, mapping=aes(x=len, y=mis))
pl3 <- pl3 + geom_smooth() +geom_point()
pl3
```


** Turning in the assignment**

* Click the "Preview" Button to generate an up-to-date html version of your notebook.  Check it to make sure you are happy with its content.
* add your .Rmd and .nb.html files to the repository and commit the changes.
* push the changes
