---
title: "R Notebook"
output: html_notebook
---

__Name:__ Mitali Singh

__Student ID:__ 999290264

For each of the exercises below, take advantage of R markdown and the notebook features.  That is include the code and have Rstudio embed your answers.  See Exercise 0 for an example


## Exercise 0 (example)
_Provide the name of your current working directory and list the files therein_
```{r}
getwd()
dir()
```

(You would then click on the > arrow in your answer code block to run it before continuing)

## Exercise 1

_Use sum() to determine the sum of numbers from 2000 to 20000.  Provide your code and the actual sum in your answer._
- 
```{r}
sum(2000,20000)
```

## Exercise 2
_In one or two sentences, describe what the above code snippet it did._

- 5 is assigned to a variable 'a' and a sequence of numbers from 2 to 20 are assigned to another variable 'b'. The command is 3rd and 4th line simply reads the content of variable a and b respectively.

## Exercise 3
_Add the contents of a and b together and place the results in a new object.  Examine the result.  Include your code.  Try using both `sum()` and `+`; do you get different results?  If so, why?_

```{r}
a <- 5 
b <- 2:20
sum <- sum(a,b)
sum
```
 This command sums up all the numbers in both variables together (2-20 and 5) and gives us one single result.
 
```{r}
sum2 <- a+b
sum2
```
Whereas the command with '+' runs a loop of adding the element of variable a (5) to the elements of variable b (2-20) one at a time. So, it adds 5 to 2 first and then 5 to 3 and so on. 

## Exercise 4
_What is the sum of the 5th through 10th element of object b?  Provide your code and the sum._
```{r}
sum (b [c(5:10)])
```


## Exercise 5
_What is the sum of the 3rd, 8th, and 10th element of b?  For both of these exercises should only need to have "b" in your code once._
```{r}
sum (b[c(3,8,10)])
```


## Exercise 6
_When extracting from a 2-dimensional object, which number specifies rows and which specifies columns?_

_What does `m[3,]` do?_

_How can you extract the 3rd, 4th and 5th columns of m together as one object?_


```{r}
m <- matrix(data=1:25, ncol=5, byrow=T)
m
m[5,5]
```
- The first number identifies row and the second number identifies column
```{r}
m[3,]
```

- ```m[3,]``` displays all the elements of row 3 (includes all columns for row 3)
```{r}
m[,3:5]
```



## Exercise 7
_What does the cbind command do?  How about rbind?_
```{r}
cbind(m,101:105)
```



```{r}
rbind(m,101:105)
```


_Create a new object "n" where the first row is a new row of numbers (your choice) and the following rows are the followed by the m matrix._

_Want more? (OPTIONAL) do the same but reverse the order of the rows from m._


## Exercise 8


```{r}
fly.worm <- read.delim("~/Downloads/fly_worm_blast.out.gz",header=F)
head(fly.worm)
summary(fly.worm)
```


```{r}
colnames(fly.worm) <-c("query_id", "subject_id", "pct_ident", "len", "mis", "gaps", "gb", "qe", "sb", "se", "E", "Score")
head(fly.worm)
summary(fly.worm)
Evalue <- fly.worm[,11]
Evalue==0
table(Evalue==0)

Percentid <- fly.worm[,3]
Percentid>50
table(Percentid>50)
table (Evalue==0&Percentid<50)

```

_How many hits have an e-value of 0?_
- 2443 hits have an e-value of 0

```{r}
table(Evalue==0)
```


_How many have hits have a percent identity > 50?_
```{r}
table(Percentid>50)
```

- 5271 hits have a percent identity >5

_Recalculate the above values but in percentage of hits rather than absolute values._

- for E_value=0: 
2443/34,874=7%

- For percent identity>50:
 5271/(29603+5271)= 15.1%
 
_How many hits have an E-value of 0 and have a percent identity less than 50?_
```{r}
table (Evalue==0&Percentid<50)
```

- 948 hits

_What is the minimum percent identity of the hits with an E-value equal to 0?_
```{r}

min(fly.worm[which(fly.worm["E"]==0),"pct_ident"])
```
- The minimum percent identity of the hits with an E-value to 0 is 24.07


## Excercise 9
_Are you surprised that sequences with relatively low percent identity can still have an E-value of 0?_
- yes

_State a hypothesis about what alignment properties might produce a zero E-value even when the percent identity is less than 50%._

- The aligned sequences could have large gaps but the amino acids that actually align could have an exact alignment of amino acids (less mismatches). This would give a low E-value and low percent identity.

_Test your hypothesis_
```{r}
allEval<- fly.worm[which(fly.worm["E"]==0),]
allEval[which(allEval["pct_ident"]<50),]
```
My hypothesis is wrong because the sequence has high number of mismatches (1508) and high umber of gaps (56). This means that the low E-value was not due to perfect alignment of the amino acids that actually aligned. 


## Excercise 10
_It sometimes is useful to create a random subset of a large data set like the fly.worm results.  Create a new dataframe `fly.worm.small` that has 1,000 rows taken at random from the larger data set._

```{r}
fly.worm[sample(nrow(fly.worm),1000),]
fly.worm.small <-fly.worm[sample(nrow(fly.worm),1000),]
```

## Exercise 11
_Use the `summary()` and `head()` functions to examine the stereotypes object._
```{r}
stereotype <-read.csv("~/Downloads/stereotypes.csv",header=T)
summary(stereotype)
head(stereotype)
```

_Record the answers to the "activity break sections" from Ciera's visualiztio tutorial here:_ 
 1) Nerd females binge drink coffee but hardly ever drink alcohol
 -
```{r}
nerds_and_metalheads <- subset(stereotype, population == "nerd" | population == "metalhead")

males.only<- subset(stereotype, gender == "male")
males.metalheads <- subset(males.only, population == "metalhead"& beer > 25& tacos > 20)
males.hipsters <- subset(males.only, population == "hipster"& beer > 25& tacos > 20)
nrow(males.metalheads)
nrow(males.hipsters)
```

2) Yes, there are 64  more metalheads who binge drink alcohol and eat >20 tacos in a week.

3) Are there more males in the metalheads who binge coffee (over 12 a week) and eat over 8 tacos a week in the metal heads or females metalheads?
```{r}
females.only<- subset(stereotype, gender == "female")
females.metalheads <- subset(females.only, population == "metalhead"& coffee > 12& tacos > 8)
males.metalheads <- subset(males.only, population == "metalhead"& coffee > 12& tacos > 8)
nrow(females.metalheads)
nrow(males.metalheads)
```
Yes, there are 35 more male metalheads than female metalheads who binge drink coffee (over 12 a week) and eat over 8 tacos a week.

4) 


## Exercise 12
_Use ggplot to explore the relationship between score ("Score") and alignment length("len").  Provide a plot that illustrates the relationship and describe the relationship in words._
```{r}
data(fly_worm_blast.out.gz)
```



## Excercise 13
_While you might expect that BLAST results with long alignments would have high scores, this is not always the case.  Form a hypothesis as to what might influence the relationship between alignment length and score.  Use ggplot to make a new plot to explore this hypothesis.  Does the plot support your hypothesis or not?  State your hypothesis, provide the code for your plot, and state your conclusion._

** Turning in the assignment**

* Click the "Preview" Button to generate an up-to-date html version of your notebook.  Check it to make sure you are happy with its content.
* add your .Rmd and .nb.html files to the repository and commit the changes.
* push the changes
